3scaleAdapter/threescale_test.go

```diff
@@ -0,0 +1,131 @@
+package threescale
+
+import (
+	"context"
+	"fmt"
+	"github.com/gogo/protobuf/types"
+	"istio.io/api/policy/v1beta1"
+	"istio.io/istio/mixer/template/authorization"
+	"net"
+	"net/http"
+	"net/http/httptest"
+	"net/url"
+	"reflect"
+	"testing"
+)
+
+func Test_buildRequestFromInstanceMsg(t *testing.T) {
+
+	expectedURLObject := &http.Request{
+		Method: "get",
+		URL: &url.URL{
+			User: &url.Userinfo{},
+			Path: "/path?query=1&query=2&query=3&user_key=12345",
+		},
+		Header: map[string][]string{"user-key": {"user_key_header"}},
+	}
+
+	instanceMsg := &authorization.InstanceMsg{
+		Name: "authorizationTest",
+		Subject: &authorization.SubjectMsg{
+			User:       "",
+			Groups:     "",
+			Properties: nil,
+		},
+		Action: &authorization.ActionMsg{
+			Namespace: "3scale",
+			Service:   "myservice",
+			Method:    "get",
+			Path:      "/path?query=1&query=2&query=3&user_key=12345",
+			Properties: map[string]*v1beta1.Value{"user-key": {
+				&v1beta1.Value_StringValue{
+					StringValue: "user_key_header",
+				},
+			}},
+		},
+	}
+
+	originalRequest := buildRequestFromInstanceMsg(instanceMsg)
+
+	if reflect.DeepEqual(expectedURLObject, originalRequest) {
+		t.Logf("the same")
+	}
+}
+
+func Test_HandleAuthorization_Ok(t *testing.T) {
+
+	ctx := context.TODO()
+
+	r := &authorization.HandleAuthorizationRequest{
+		Instance: &authorization.InstanceMsg{
+			Subject: &authorization.SubjectMsg{},
+			Action:  &authorization.ActionMsg{},
+		},
+		AdapterConfig: &types.Any{},
+		DedupId:       "",
+	}
+
+	ts := httptest.NewUnstartedServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		w.WriteHeader(http.StatusOK)
+		fmt.Fprintf(w, "Ok")
+	}))
+	l, _ := net.Listen("tcp", "127.0.0.1:8090")
+	ts.Listener = l
+	ts.Start()
+	defer ts.Close()
+	result, err := (*Threescale).HandleAuthorization(&Threescale{}, ctx, r)
+	if err != nil {
+		t.Errorf("Fail %#v", err)
+	}
+
+	if result.Status.Code != 0 {
+		t.Errorf("Result different from expected %#v", result.Status.Code)
+	}
+
+}
+
+func Test_HandleAuthorization_Failing(t *testing.T) {
+
+	ctx := context.TODO()
+
+	r := &authorization.HandleAuthorizationRequest{
+		Instance: &authorization.InstanceMsg{
+			Subject: &authorization.SubjectMsg{},
+			Action:  &authorization.ActionMsg{},
+		},
+		AdapterConfig: &types.Any{},
+		DedupId:       "",
+	}
+
+	ts := httptest.NewUnstartedServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		w.WriteHeader(http.StatusForbidden)
+		fmt.Fprintf(w, "Forbidden")
+	}))
+	l, _ := net.Listen("tcp", "127.0.0.1:8090")
+	ts.Listener = l
+	ts.Start()
+	defer ts.Close()
+
+	result, err := (*Threescale).HandleAuthorization(&Threescale{}, ctx, r)
+	if err != nil {
+		t.Errorf("Fail %#v", err)
+	}
+
+	if result.Status.Code != 7 {
+		t.Errorf("Result different from expected %#v", result.Status.Code)
+	}
+	ts.Close()
+}
```

Comment:
I would table-ize (new word) the above tests as discussed. Also it would be good to allow injection of http client into `Threescale` struct so we can customise it and make these tests easier to write.